### ISSUE TITLE
Vehicle-client restrictions

### ISSUE THREAD

**Issue by N-Wouda** [PyVRP/PyVRP#329]
#### Describe the solution you'd like

Currently, all vehicles can be used to visit all clients. This might be OK in static problem settings like the ones we currently support, but falls apart when e.g. re-optimizing routes that are already underway: the deliveries have already been loaded onto the truck, and the clients those deliveries go to cannot be serviced by another vehicle.

I *might* need this for a case I'm solving, but I can also solve my particular instance in a less efficient manner by tweaking the time windows. But that does not hold in general.

#### Describe alternatives you have considered

There aren't any, really. This is currently unsupported.

#### Additional context

We could support this by restricting the vehicle types to which a client may be assigned. There can only be one vehicle of such a type (how do we enforce that cleanly?). Such a client should be assigned to that vehicle type, but I am not sure if we should require that or deal with it via soft penalties like load or time window violations.

**Comment by N-Wouda** [comment]
I think soft violations and a new "swap vehicle" route operator. We'll need such a route operator anyway for #188, and likely for #305 as well.

**Comment by leonlan** [comment]
The VRP with vehicle-client assignment restrictions is called "site-dependent VRP" in the literature. Instead of soft violations, I think you could also do this with a bitmaps.

**Comment by leonlan** [comment]
Just a note that the MDVRPTW instances from Vidal et al. (2013) are also used for the site-dependent VRPTW.

**Comment by lstegner** [comment]
Hello! I am wondering if there is a plan/timeline for site-dependent VRP support in the near future, or if there's some advice/guidance to get a "quick and dirty" solution up and running for some small scale testing for a research project I'm working on.

**Comment by N-Wouda** [comment]
Hi @lstegner! 👋 A realistic timeline for this is before the summer, say in June. We'll add support for site-dependent VRP via #475, which will be part of PyVRP 0.9.0 (the next release). That'll likely take around 2-3 months, as I am currently quite busy finishing up my dissertation.

**Comment by lstegner** [comment]
Hi @N-Wouda! Thanks so much for the quick update. Good luck with your dissertation, and looking forward to the next release 😄

**Comment by N-Wouda** [comment]
Hi @lstegner, with #475 now completed, it is possible to model vehicle-client restrictions using a workaround with profiles. In particular, by making a routing profile per vehicle type, you could model the vehicle-client restrictions as follows:
```python
from pyvrp import Model

m = Model()

# here add your depots and clients

prof1 = m.add_profile()
veh_type1 = m.add_vehicle_type(..., profile=prof1)

prof2 = m.add_profile()
veh_type2 = m.add_vehicle_type(..., profile=prof2)

for frm in m.locations:
  for to in m.locations:    
    dist1 = ...
    m.add_edge(frm, to, dist=dist1, profile=prof1)
 
    dist2 = ...
    m.add_edge(frm, to, dist=dist2, profile=prof2)
```
where `dist1 = ...` and `dist2 = ...` should be made really big if `to` cannot be visited by the first or second profile (and thus first and second vehicle type), respectively. We will probably improve our support for vehicle-client restrictions in the coming releases, but for now this workaround should be sufficient to get started. Hope it's all clear!

Note that this has not yet been released to PyPI, so you cannot get it via regular `pip install pyvrp` yet. If you want to use a precompiled binary, please download the appropriate wheel for your platform from one of our CD runs, [here](https://github.com/PyVRP/PyVRP/actions/workflows/CD.yml), and install that using pip.

**Comment by N-Wouda** [comment]
Benchmarks and potential performance improvements needed for SDVRP are in scope for v0.10.0. The benchmark instances and parsing routines will already be part of v0.9.0 via #617.

**Comment by leonlan** [comment]
We currently do not have SDVRPTW benchmarks yet because the performance on large instances is subpar. I think this is because of SREX, which does not seem to preserve vehicle-client restrictions from parents to offspring. I won't spend time on improving HGS futher because we are moving towards ILS.

Here are the results from earlier benchmarks. For each instance, the average gap w.r.t. BKS is presented over 10 runs. Time limits were 300 seconds for Cordeau, 3600 seconds for Vidal. Observations

<details>
<summary>Details</summary>

```
# Cordeau 2001
PR01    0.000
PR02    0.000
PR03    0.162
PR04    0.134
PR05    0.460
PR06    0.502
PR07    0.000
PR08    0.136
PR09    0.314
PR10    0.524
PR11    0.000
PR12   -0.000
PR13    0.120
PR14    0.120
PR15    0.221
PR16    0.575
PR17    0.000
PR18    0.143
PR19    0.273
PR20    0.566
PR21    8.252
PR22    5.281
PR23    8.816
PR24    4.745

# Vidal 2013
PR11A   -0.191
PR11B   -0.802
PR12A    0.593
PR12B   -0.005
PR13A    1.790
PR13B    0.453
PR14A    1.991
PR14B    2.612
PR15A    4.406
PR15B    3.586
PR16A    5.951
PR16B    3.795
PR17A   -0.025
PR17B   -0.082
PR18A   -0.020
PR18B    0.093
PR19A    2.958
PR19B    2.644
PR20A    5.065
PR20B    4.337
PR21A    0.024
PR21B   -0.538
PR22A    2.492
PR22B    0.576
PR23A    4.758
PR23B    3.983
PR24A    7.522
PR24B    6.102
```

</details>

Observations
- On small instances (<400 clients) PyVRP performs well.
- On larger instances, the performance is quite lackluster with gaps between 1-8%.

**Comment by N-Wouda** [comment]
We can (ab)use multiple loads for this, I think. Let's try another benchmark run once we have completed/figured out #645.

**Comment by leonlan** [comment]
Tried thinking of addressing vehicle-client restrictions with multiple load dimensions (#645), but it does not seem practical. Our assumption was that the allowed client groups were mutually exclusive (like in CluVRP) but that's not the case: vehicle type 1 could serve clients 1 and 2, while vehicle type 2 could serve clients 1 and 3. The transformation of one load dimension per client group does not work in this case.

**Comment by leonlan** [comment]
@N-Wouda I think this issue can be closed, right?

- We have an example that showcases how to implement vehicle-client restrictions.
- Current version already supports reading SDVRPTW instances (but performance is subpar on large instances).
- SDVRPTW instances are added in https://github.com/PyVRP/Instances/pull/11.
- We'll skip benchmarking for now. Possibly won't do the benchmarking again in view of #503.

**Comment by N-Wouda** [comment]
@leonlan agree. We might want to do a benchmark for 0.12 on this for good measure, but I think current performance on this particular variant is already sufficiently good, too.
